pysys/process/plat-win32/helper.py

```diff
@@ -112,11 +112,79 @@ def _writeStdin(self, data):
 			else:
 				win32file.WriteFile(self.__stdin, data, None)
 
-	def __quotePath(self, input):
-		"""Private method to escape a windows path according to documented guidelines for this OS.
+	def __quoteCommand(self, input):
+		"""Private method to quote a command (argv[0]) "correctly" for Windows.
+
+		The returned value can be used as the start of a command line, with subsequent
+		arguments quoted using __quoteArgument() and added to the command line with
+		intervening spaces.
+		
+		The quoted command will be handled correctly by the standard Windows command
+		line parsers (CommandLineToArgvW and parse_cmdline), unless the input includes
+		double quotes, control characters, or whitespace other than spaces.  The
+		behaviour of the standard parsers is different in those cases, and given that
+		there is no way of knowing which parser the command will actually use, no
+		attempt is made to deal with these differences and the results are undefined.
+		"""
+		return '\"%s\"' % input if ' ' in input else input
+
+	def __quoteArgument(self, input):
+		"""Private method to quote and escape a command line argument correctly for Windows.
+
+		The returned value can be added to the end of a command line, with an intervening
+		space, and will be treated as a separate argument by the standard Windows command
+		line parsers (CommandLineToArgvW and parse_cmdline).  Double quotes, whitespace
+		and backslashes in the argument will be preserved for the parser to see them.
+		
+		Windows' quoting and escaping rules are somewhat complex and the implementation
+		of this method was derived from a few different sources:
+		https://docs.microsoft.com/en-us/previous-versions/17w5ykft(v=vs.85)
+		http://www.windowsinspired.com/the-correct-way-to-quote-command-line-arguments/
+		http://www.windowsinspired.com/understanding-the-command-line-string-and-arguments-received-by-a-windows-program/
+		http://www.windowsinspired.com/how-a-windows-programs-splits-its-command-line-into-individual-arguments/
+		https://daviddeley.com/autohotkey/parameters/parameters.htm
+		This method tries to avoid any areas of different behaviour between the two
+		standard parsers, in particular the undocumented rules around handling of
+		consecutive unescaped double quote characters.
+		"""
+		output = ""
```

Comment:
This _could_ be premature optimization, but we don't have the data to be sure and this startProcess code is executed a LOT... so how about reverting to a trivial but safe implementation if input doesn't contain space/tab/doublequote chars (or backslash??)?

Reply:
Hmm... I would expect the cost of actually starting the process to dominate (I remember that being relatively expensive on Windows) but I can whip up a quick performance test to compare with the old implementation.  If it's worthwhile then falling back to a dumb implementation for the simple (and likely most common) case seems reasonable.  Stay tuned.

Reply:
Well this was probably a bit unnecessary, but I wrote a performance test (PySys_internal_perf_003) and once I had that it seemed only reasonable to make some optimisations... Anyway, full results below but in summary all the easy cases are now faster than they were, and the not-so-easy ones (basically anything with double-quotes _within_ an element of argv) are significantly slower.  I don't think those are particularly common cases, and even on my ancient laptop I could still generate around 6,000+ command lines/second in the worst case, so it shouldn't be a big problem in any reasonable real-world situation.

Results below are comparing several versions of the new algorithm to the "legacy" implementation so lower speedup is actually better.  "unoptimised" is my first attempt (all round bad performance), "shortcircuit" special-cases the easy stuff, and "stringjoin" uses `join()` instead of string concatenation - the last one shows the best overall improvement and is what's in the PR.  BTW the performance results comparison stuff is really cool :-)
```
Performance comparison summary between the following runs (the differing "run details" are shown):
- 2022-06-12_16.59.40.win.unoptimised (8 result keys, 10 samples/result):
     vcsCommit=c93aa462, startTime=2022-06-12 16:59:40
- 2022-06-12_17.28.28.win.shortcircuit (8 result keys, 10 samples/result):
     vcsCommit=c93aa462, startTime=2022-06-12 17:28:28
- 2022-06-12_22.27.58.win.stringjoin (8 result keys, 10 samples/result):
     vcsCommit=95fc7a39, startTime=2022-06-12 22:27:58
- 2022-06-12_15.51.24.win.legacy (8 result keys, 10 samples/result):
     vcsCommit=c93aa462, xargs=useLegacyAlgorithm-True, startTime=2022-06-12 15:51:24

Comparison results:
  where fromRun->toRun format is: (%improvement) = (speedup ratio) = (sigmas/stdDevs where change above +/- 2.0 gives 95% probability it's significant; only significant results are colored)

Command line generation rate (1: command / no args) from PySys_internal_perf_003
 Mean from this run = 684,124 /s with stdDev=231,645 (33.9% of mean)
   2022-06-12_16.59.40.win.unoptimised->2022-06-12_15.51.24.win.legacy: -48.2% = 0.5183x speedup = -1.4 sigmas
  2022-06-12_17.28.28.win.shortcircuit->2022-06-12_15.51.24.win.legacy: -37.3% = 0.6268x speedup = -1.1 sigmas
    2022-06-12_22.27.58.win.stringjoin->2022-06-12_15.51.24.win.legacy: -39.9% = 0.6007x speedup = -1.2 sigmas

Command line generation rate (2: command with path / no args) from PySys_internal_perf_003
 Mean from this run = 665,764 /s with stdDev=211,251 (31.7% of mean)
   2022-06-12_16.59.40.win.unoptimised->2022-06-12_15.51.24.win.legacy: -48.9% = 0.5108x speedup = -1.5 sigmas
  2022-06-12_17.28.28.win.shortcircuit->2022-06-12_15.51.24.win.legacy: -37.2% = 0.6281x speedup = -1.2 sigmas
    2022-06-12_22.27.58.win.stringjoin->2022-06-12_15.51.24.win.legacy: -39.8% = 0.6017x speedup = -1.3 sigmas

Command line generation rate (3: command with spaces / no args) from PySys_internal_perf_003
 Mean from this run = 658,374 /s with stdDev=211,629 (32.1% of mean)
   2022-06-12_16.59.40.win.unoptimised->2022-06-12_15.51.24.win.legacy: -29.1% = 0.7091x speedup = -0.9 sigmas
  2022-06-12_17.28.28.win.shortcircuit->2022-06-12_15.51.24.win.legacy: -13.0% = 0.8705x speedup = -0.4 sigmas
    2022-06-12_22.27.58.win.stringjoin->2022-06-12_15.51.24.win.legacy: -20.0% = 0.8003x speedup = -0.6 sigmas

Command line generation rate (4: simple args) from PySys_internal_perf_003
 Mean from this run = 79,617 /s with stdDev=25,669 (32.2% of mean)
   2022-06-12_16.59.40.win.unoptimised->2022-06-12_15.51.24.win.legacy: +488.4% = 5.884x speedup = +15.1 sigmas
  2022-06-12_17.28.28.win.shortcircuit->2022-06-12_15.51.24.win.legacy:  -7.7% = 0.9234x speedup = -0.2 sigmas
    2022-06-12_22.27.58.win.stringjoin->2022-06-12_15.51.24.win.legacy: -46.6% = 0.5342x speedup = -1.4 sigmas

Command line generation rate (5: args with path) from PySys_internal_perf_003
 Mean from this run = 76,030 /s with stdDev=24,456 (32.2% of mean)
   2022-06-12_16.59.40.win.unoptimised->2022-06-12_15.51.24.win.legacy: +418.6% = 5.186x speedup = +13.0 sigmas
  2022-06-12_17.28.28.win.shortcircuit->2022-06-12_15.51.24.win.legacy: -11.5% = 0.8848x speedup = -0.4 sigmas
    2022-06-12_22.27.58.win.stringjoin->2022-06-12_15.51.24.win.legacy: -50.0% = 0.4998x speedup = -1.6 sigmas

Command line generation rate (6: args with spaces) from PySys_internal_perf_003
 Mean from this run = 79,786 /s with stdDev=24,010 (30.1% of mean)
   2022-06-12_16.59.40.win.unoptimised->2022-06-12_15.51.24.win.legacy: +527.5% = 6.275x speedup = +17.5 sigmas
  2022-06-12_17.28.28.win.shortcircuit->2022-06-12_15.51.24.win.legacy: +17.1% = 1.171x speedup = +0.6 sigmas
    2022-06-12_22.27.58.win.stringjoin->2022-06-12_15.51.24.win.legacy: -29.6% = 0.7039x speedup = -1.0 sigmas

Command line generation rate (7: args with quotes) from PySys_internal_perf_003
 Mean from this run = 66,775 /s with stdDev=18,397 (27.6% of mean)
   2022-06-12_16.59.40.win.unoptimised->2022-06-12_15.51.24.win.legacy: +408.0% = 5.08x speedup = +14.8 sigmas
  2022-06-12_17.28.28.win.shortcircuit->2022-06-12_15.51.24.win.legacy: +528.2% = 6.282x speedup = +19.2 sigmas
    2022-06-12_22.27.58.win.stringjoin->2022-06-12_15.51.24.win.legacy: +413.5% = 5.135x speedup = +15.0 sigmas

Command line generation rate (8: complex args) from PySys_internal_perf_003
 Mean from this run = 52,031 /s with stdDev=14,779 (28.4% of mean)
   2022-06-12_16.59.40.win.unoptimised->2022-06-12_15.51.24.win.legacy: +630.2% = 7.302x speedup = +22.2 sigmas
  2022-06-12_17.28.28.win.shortcircuit->2022-06-12_15.51.24.win.legacy: +806.0% = 9.06x speedup = +28.4 sigmas
    2022-06-12_22.27.58.win.stringjoin->2022-06-12_15.51.24.win.legacy: +602.7% = 7.027x speedup = +21.2 sigmas
```

Reply:
Not sure if you read the previous set of performance results but if not, don't bother.  My laptop was suffering from some thermal problems and throttling the CPU under load, so I wouldn't trust any those numbers.  Anyway that problem is fixed and I have re-run the the performance test.  The runs compared below are:
1. `legacy` - the old, simple, but not correct in all cases algorithm
2. `optimised1` - the new algorithm with some easy/common cases short-circuited and all the string concatenation replaced with `join()`s
3. `optimised2` - small tweaks to make sure we get the string length and scan for whitespace exactly once

In summary, compared to the legacy algorithm:
* Cases with no whitespace and no inline double quotes are a bit faster
* Cases with whitespace but no inline double quotes are a bit slower
* Anything with inline double quotes is a _lot_ slower, but still doing ~9,000 command lines/sec on a relatively ancient laptop in the worst case, so I wouldn't expect it to be a problem in the real world, especially as I don't expect this to be a common use case anyway

What do you think?  My view would be that any performance hit in a realistic scenario is going to be lost in the noise...

```
Performance comparison summary between the following runs (the differing "run details" are shown):
- 2022-06-18_10.09.40.win.legacy (8 result keys, 10 samples/result):
     xargs=useLegacyAlgorithm-True, startTime=2022-06-18 10:09:40
- 2022-06-18_10.24.03.win.optimised1 (8 result keys, 10 samples/result):
     startTime=2022-06-18 10:24:03
- 2022-06-18_10.49.40.win.optimised2 (8 result keys, 10 samples/result):
     startTime=2022-06-18 10:49:40

Comparison results:
  where fromRun->toRun format is: (%improvement) = (speedup ratio) = (sigmas/stdDevs where change above +/- 2.0 gives 95% probability it's significant; only significant results are colored)

Command line generation rate (1: command / no args) from PySys_internal_perf_003
 Mean from this run = 1,344,825 /s with stdDev=118,765 (8.8% of mean)
      2022-06-18_10.09.40.win.legacy->2022-06-18_10.49.40.win.optimised2:  +8.0% = 1.08x speedup = +0.9 sigmas
  2022-06-18_10.24.03.win.optimised1->2022-06-18_10.49.40.win.optimised2:  -1.6% = 0.9838x speedup = -0.2 sigmas

Command line generation rate (2: command with path / no args) from PySys_internal_perf_003
 Mean from this run = 1,308,619 /s with stdDev=37,703 (2.9% of mean)
      2022-06-18_10.09.40.win.legacy->2022-06-18_10.49.40.win.optimised2:  +7.2% = 1.072x speedup = +1.0 sigmas
  2022-06-18_10.24.03.win.optimised1->2022-06-18_10.49.40.win.optimised2:  -0.9% = 0.9914x speedup = -0.3 sigmas

Command line generation rate (3: command with spaces / no args) from PySys_internal_perf_003
 Mean from this run = 996,372 /s with stdDev=17,481 (1.8% of mean)
      2022-06-18_10.09.40.win.legacy->2022-06-18_10.49.40.win.optimised2: -17.6% = 0.8239x speedup = -2.0 sigmas
  2022-06-18_10.24.03.win.optimised1->2022-06-18_10.49.40.win.optimised2:  -1.3% = 0.9872x speedup = -0.6 sigmas

Command line generation rate (4: simple args) from PySys_internal_perf_003
 Mean from this run = 174,074 /s with stdDev=2,914 (1.7% of mean)
      2022-06-18_10.09.40.win.legacy->2022-06-18_10.49.40.win.optimised2: +24.1% = 1.241x speedup = +4.9 sigmas
  2022-06-18_10.24.03.win.optimised1->2022-06-18_10.49.40.win.optimised2:  -2.4% = 0.9758x speedup = -0.3 sigmas

Command line generation rate (5: args with path) from PySys_internal_perf_003
 Mean from this run = 178,622 /s with stdDev=2,242 (1.3% of mean)
      2022-06-18_10.09.40.win.legacy->2022-06-18_10.49.40.win.optimised2: +30.4% = 1.304x speedup = +8.6 sigmas
  2022-06-18_10.24.03.win.optimised1->2022-06-18_10.49.40.win.optimised2:  +8.1% = 1.081x speedup = +0.5 sigmas

Command line generation rate (6: args with spaces) from PySys_internal_perf_003
 Mean from this run = 112,490 /s with stdDev=2,422 (2.2% of mean)
      2022-06-18_10.09.40.win.legacy->2022-06-18_10.49.40.win.optimised2: -20.9% = 0.791x speedup = -9.7 sigmas
  2022-06-18_10.24.03.win.optimised1->2022-06-18_10.49.40.win.optimised2:  -0.4% = 0.9963x speedup = -0.0 sigmas

Command line generation rate (7: args with quotes) from PySys_internal_perf_003
 Mean from this run = 15,974 /s with stdDev=170.3 (1.1% of mean)
      2022-06-18_10.09.40.win.legacy->2022-06-18_10.49.40.win.optimised2: -86.5% = 0.1355x speedup = -7.1 sigmas
  2022-06-18_10.24.03.win.optimised1->2022-06-18_10.49.40.win.optimised2:  +6.1% = 1.061x speedup = +0.6 sigmas

Command line generation rate (8: complex args) from PySys_internal_perf_003
 Mean from this run = 8,996 /s with stdDev=335.9 (3.7% of mean)
      2022-06-18_10.09.40.win.legacy->2022-06-18_10.49.40.win.optimised2: -90.5% = 0.09485x speedup = -19.0 sigmas
  2022-06-18_10.24.03.win.optimised1->2022-06-18_10.49.40.win.optimised2:  +4.8% = 1.048x speedup = +0.4 sigmas
```

Reply:
so it's just a bit slower for the common case? sounds ok then! 

i'm a little more cautious about performance of critical-path stuff (like starting processes) the more CPUs we get in modern machines given the GIL problem. i.e. I don't care about the time taken for a single test, but want to keep scalability good when running lots of tests in parallel